builtin/nomad/jobspec/releaser.go

```diff
@@ -0,0 +1,547 @@
+	sdk "github.com/hashicorp/waypoint-plugin-sdk/proto/gen"
+	"github.com/hashicorp/waypoint-plugin-sdk/terminal"
+	"github.com/hashicorp/waypoint/builtin/nomad"
+)
+
+// Releaser is the ReleaseManager implementation for Nomad.
+type Releaser struct {
+	p      *Platform
+	config ReleaserConfig
+}
+
+// Config implements Configurable
+func (r *Releaser) Config() (interface{}, error) {
+	return &r.config, nil
+}
+
+// ReleaseFunc implements component.ReleaseManager
+func (r *Releaser) ReleaseFunc() interface{} {
+	return r.Release
+}
+
+// DestroyFunc implements component.Destroyer
+func (r *Releaser) DestroyFunc() interface{} {
+	return r.Destroy
+}
+
+// StatusFunc implements component.Status
+func (r *Releaser) StatusFunc() interface{} {
+	return r.Status
+}
+
+func (r *Releaser) resourceManager(log hclog.Logger, dcr *component.DeclaredResourcesResp) *resource.Manager {
+	return resource.NewManager(
+		resource.WithLogger(log.Named("resource_manager")),
+		resource.WithValueProvider(r.getNomadClient),
+		resource.WithDeclaredResourcesResp(dcr),
+		resource.WithResource(resource.NewResource(
+			resource.WithName(rmResourceJobName),
+			resource.WithState(&Resource_Job{}),
+			resource.WithCreate(r.resourceJobCreate),
+			resource.WithDestroy(r.resourceJobDestroy),
+			resource.WithStatus(r.resourceJobStatus),
+			resource.WithPlatform("nomad"),
+			resource.WithCategoryDisplayHint(sdk.ResourceCategoryDisplayHint_INSTANCE_MANAGER),
+		)),
+	)
+}
+
+// getNomadClient provides
+// the client connection used by resources to interact with Nomad.
+func (r *Releaser) getNomadClient() (*nomadClient, error) {
+	// Get our client
+	client, err := api.NewClient(api.DefaultConfig())
+	if err != nil {
+		return nil, err
+	}
+	return &nomadClient{
+		NomadClient: client,
+	}, nil
+}
+
+func (r *Releaser) resourceJobCreate(
+	ctx context.Context,
+	log hclog.Logger,
+	target *Deployment,
+	result *Release,
+	state *Resource_Job,
+	client *nomadClient,
+	st terminal.Status,
+) error {
+	// Set up clients
+	jobClient := client.NomadClient.Jobs()
+	deploymentClient := client.NomadClient.Deployments()
+
+	st.Update("Getting job...")
+	jobs, _, err := jobClient.PrefixList(target.Name)
+	if err != nil {
+		return status.Errorf(codes.Aborted, "Unable to fetch Nomad jobs: %s", err.Error())
+	} else if target.Name != jobs[0].ID {
+		return status.Errorf(codes.Aborted, "Job not found: %s", err.Error())
+	}
+
+	q := &api.QueryOptions{
+		Namespace: jobs[0].JobSummary.Namespace,
+	}
+
+	job, _, err := jobClient.Info(jobs[0].ID, q)
+
+	if err != nil {
+		return status.Errorf(codes.Aborted, "Unable to fetch Nomad job: %s", err.Error())
+	}
+
+	st.Update("Getting latest deployments for job")
+	deploy, _, err := jobClient.LatestDeployment(*job.ID, q)
+	if err != nil {
+		return status.Errorf(codes.Aborted, "Unable to fetch latest deployment for Nomad job: %s", err.Error())
+	} else if deploy == nil {
+		st.Update("No active deployment for Nomad job")
+		return err
+	}
```

Comment:
Because we have the deployment at this point, we might want to consider checking the initial deployment status before we go rummaging through the task groups. The deployment might have already failed, completed, been marked successful, etc.

Also, "latest deployment" means literally that and not necessarily the one you're expecting (this is admittedly a bit of a gotcha in the Nomad API!). When you submit a job:

* (1) The job is forwarded to the leader, and stored in the Nomad state store. At the same time, an evaluation is created.
* (2) A scheduler worker (not necessarily on the leader) dequeues the evaluation and creates a plan with the deployment. The plan is sent to the leader to be "applied".

So imagine if we submit a job via the API and then immediately hit `LatestDeployment` before the new deployment is created. We could end up with the deployment for the _previous_ version of the job, which would be confusing because we'd try to promote a possibly already-promoted deployment. So we'll want to match the version for this deployment against the job version we've just submitted.

Reply:
@tgross I caught up to this comment after my response to your remarks on the status check, and I agree a similar "early exit" strategy could be applied here if the deployment fails/is unhealthy.

And thank you for the tip on the gotcha! I'll add in a check to verify that the job version returned by the job `Info` call matches up with the job version returned by the `LatestDeployment`.